ouroboros-network-protocols/src/Ouroboros/Network/Protocol/ObjectDiffusion/Type.hs

```diff
@@ -0,0 +1,359 @@
+pattern ReflOutboundAgency = ReflClientAgency
+type ReflOutboundAgency = 'ReflClientAgency
+
+pattern ReflInboundAgency :: ReflRelativeAgency ServerAgency (r :: RelativeAgency) (r :: RelativeAgency)
+pattern ReflInboundAgency = ReflServerAgency
+type ReflInboundAgency = 'ReflServerAgency
+
+-- | The kind of the object diffusion protocol, and the types of the states in
+-- the protocol state machine.
+--
+-- We describe this protocol using the label \"inbound\" for the peer that is
+-- receiving objects, and \"inbound\" for the one sending them.
+type ObjectDiffusion :: Agency -> Type -> Type -> Type
+data ObjectDiffusion initAgency objectId object where
+  -- | Initial protocol message.
+  StInit :: ObjectDiffusion initAgency objectId object
+  -- | The inbound node has agency; it can either terminate, ask for object
+  -- identifiers or ask for objects.
+  --
+  -- There is no timeout in this state.
+  StIdle :: ObjectDiffusion initAgency objectId object
+  -- | The outbound node has agency; it must reply with a list of object
+  -- identifiers that it wishes to submit.
+  --
+  -- There are two sub-states for this, for blocking and non-blocking cases.
+  StObjectIds :: StBlockingStyle -> ObjectDiffusion initAgency objectId object
+  -- | The outbound node has agency; it must reply with the list of
+  -- objects.
+  StObjects :: ObjectDiffusion initAgency objectId object
+  -- | Nobody has agency; termination state.
+  StDone :: ObjectDiffusion initAgency objectId object
+
+instance
+  ( ShowProxy objectId,
+    ShowProxy object
+  ) =>
+  ShowProxy (ObjectDiffusion objectId object)
+  where
+  showProxy _ =
+    concat
+      [ "ObjectDiffusion ",
+        showProxy (Proxy :: Proxy objectId),
+        " ",
+        showProxy (Proxy :: Proxy object)
+      ]
+
+instance ShowProxy (StIdle :: ObjectDiffusion initAgency objectId object) where
+  showProxy _ = "StIdle"
+
+type SingObjectDiffusion ::
+  ObjectDiffusion initAgency objectId object ->
+  Type
+data SingObjectDiffusion k where
+  SingInit :: SingObjectDiffusion StInit
+  SingIdle :: SingObjectDiffusion StIdle
+  SingObjectIds ::
+    SingBlockingStyle stBlocking ->
+    SingObjectDiffusion (StObjectIds stBlocking)
+  SingObjects :: SingObjectDiffusion StObjects
+  SingDone :: SingObjectDiffusion StDone
+
+deriving instance Show (SingObjectDiffusion k)
+
+instance StateTokenI StInit where stateToken = SingInit
+
+instance StateTokenI StIdle where stateToken = SingIdle
+
+instance
+  (SingI stBlocking) =>
+  StateTokenI (StObjectIds stBlocking)
+  where
+  stateToken = SingObjectIds sing
+
+instance StateTokenI StObjects where stateToken = SingObjects
+
+instance StateTokenI StDone where stateToken = SingDone
+
+data StBlockingStyle where
+  -- | In this sub-state the reply need not be prompt. There is no timeout.
+  StBlocking :: StBlockingStyle
+  -- | In this state the peer must reply. There is a timeout.
+  StNonBlocking :: StBlockingStyle
```

Comment:
Why not import this type from `TxSubmission2/Type`? It can also be exported here.